Source/ARTRealtimePresence.m

```diff
@@ -780,85 +778,110 @@ - (void)reenterInternalMembers {
     return _internalMembers;
 }
 
-- (BOOL)add:(ARTPresenceMessage *)message {
-    ARTPresenceMessage *latest = [_members objectForKey:message.memberKey];
-    if ([message isNewerThan:latest]) {
-        ARTPresenceMessage *messageCopy = [message copy];
+- (void)processMember:(ARTPresenceMessage *)message {
+    ARTPresenceMessage *messageCopy = [message copy];
+    // Internal member
+    if ([message.connectionId isEqualToString:self.connectionId]) {
         switch (message.action) {
             case ARTPresenceEnter:
             case ARTPresenceUpdate:
-                messageCopy.action = ARTPresencePresent;
-                // intentional fallthrough
             case ARTPresencePresent:
-                [self internalAdd:messageCopy];
+                messageCopy.action = ARTPresencePresent;
+                [self addInternalMember:messageCopy];
                 break;
             case ARTPresenceLeave:
-                [self internalRemove:messageCopy];
+                if (!message.isSynthesized) {
+                    [self removeInternalMember:messageCopy];
+                }
                 break;
             default:
                 break;
         }
-        return YES;
     }
-    ARTLogDebug(_logger, @"Presence member \"%@\" with action %@ has been ignored", message.memberKey, ARTPresenceActionToStr(message.action));
-    latest.syncSessionId = _syncSessionId;
-    return NO;
-}
+    
+    BOOL memberUpdated = false;
+    switch (message.action) {
+        case ARTPresenceEnter:
+        case ARTPresenceUpdate:
+        case ARTPresencePresent:
+            messageCopy.action = ARTPresencePresent;
+            memberUpdated = [self addMember:messageCopy];
+            break;
+        case ARTPresenceLeave:
+            if (self.syncInProgress) {
+                messageCopy.action = ARTPresenceAbsent; // RTP2f
+                memberUpdated = [self addMember:messageCopy];
+            } else {
+                memberUpdated = [self removeMember:messageCopy];
+            }
+            break;
+        default:
+            break;
+    }
 
-- (void)internalAdd:(ARTPresenceMessage *)message {
-    [self internalAdd:message withSessionId:_syncSessionId];
+    if (memberUpdated) {
```

Comment:
```suggestion
    if (memberUpdated) {
        [_beforeSyncMembers removeObjectForKey:message.memberKey]; // RTP19
```

Reply:
See https://github.com/ably/ably-go/pull/637/files

Reply:
With current implementation, even if the member is not updated, it will remove member from `_beforeSyncMembers`, hence it won't throw `leave` event at the end of the sync.

Reply:
That's why it's added instead of removed upon `ARTPresenceLeave` (See RTP2f).

Reply:
`_beforeSyncMembers` is only for tracking members that didn't receive any update at all. For them `ARTPresenceLeave` will be generated in `didRemovedMemberNoLongerPresent:`.

Reply:
I will raise a new conversation for this 👍